File: python/1043. Partition Array for Maximum Sum.py
class Solution(object):
    def maxSumAfterPartitioning(self, A, K):
        """
        :type A: List[int]
        :type K: int
        :rtype: int
        """
        n = len(A)
        dp = [0] * (n+1)
        for i in range(n+1):
            max_val = dp[i]
            j = i-1
            while j>=0 and (i-j)<=K:
                max_val = max(max_val, A[j])
                dp[i] = max(dp[i], dp[j] + max_val*(i-j))
                j -= 1
        return dp[n]

# dp[i] 表示前i个元素，被分隔成长度最多为K的连续子数组的最大值
# 递推公式: dp[i] = max(dp[i], dp[j] + max_val*(i-j)), i-k <= j <= i-1
# time: O(N^2), space: O(N)

File: go/src/leetcode/1043/1043. Partition Array for Maximum Sum.go
package main

import "fmt"

func maxSumAfterPartitioning(arr []int, k int) int {
	n := len(arr)
	f := make([]int, n+1)
	for i := range arr {
		for j, mx := i, 0; j >= 0 && j > i-k; j-- {
			mx = max(mx, arr[j])
			f[i+1] = max(f[i+1], f[j]+(i-j+1)*mx)
		}
	}
	fmt.Println(f)
	return f[n]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(maxSumAfterPartitioning([]int{1, 15, 7, 9, 2, 5, 10}, 3))
}
